File: python-fastapi/app/routes/patient.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request, Response
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.fhir_mapper import fhir_to_patient_fields, patient_to_fhir, search_bundle
from app.models.patient import PatientModel

router = APIRouter(prefix="/fhir/Patient", tags=["Patient"])
FHIR_JSON = "application/fhir+json"


@router.get("/{patient_id}")
async def read_patient(patient_id: UUID, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(PatientModel).where(PatientModel.id == patient_id))
    patient = result.scalars().first()
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")
    return Response(
        content=_json_dumps(patient_to_fhir(patient)),
        media_type=FHIR_JSON,
    )


@router.get("")
async def search_patients(
    identifier: str | None = None, db: AsyncSession = Depends(get_db)
):
    if identifier:
        result = await db.execute(
            select(PatientModel).where(PatientModel.mrn == identifier)
        )
    else:
        result = await db.execute(select(PatientModel))
    patients = result.scalars().all()
    fhir_resources = [patient_to_fhir(p) for p in patients]
    return Response(
        content=_json_dumps(search_bundle(fhir_resources)),
        media_type=FHIR_JSON,
    )


@router.post("", status_code=201)
async def create_patient(request: Request, db: AsyncSession = Depends(get_db)):
    body = await request.json()
    fields = fhir_to_patient_fields(body)
    patient = PatientModel(**fields)
    db.add(patient)
    await db.commit()
    await db.refresh(patient)
    return Response(
        content=_json_dumps(patient_to_fhir(patient)),
        media_type=FHIR_JSON,
        headers={"Location": f"/fhir/Patient/{patient.id}"},
    )


@router.put("/{patient_id}")
async def update_patient(
    patient_id: UUID, request: Request, db: AsyncSession = Depends(get_db)
):
    result = await db.execute(select(PatientModel).where(PatientModel.id == patient_id))
    patient = result.scalars().first()
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")

    body = await request.json()
    fields = fhir_to_patient_fields(body)
    for key, value in fields.items():
        setattr(patient, key, value)
    await db.commit()
    await db.refresh(patient)
    return Response(
        content=_json_dumps(patient_to_fhir(patient)),
        media_type=FHIR_JSON,
    )


def _json_dumps(obj) -> str:
    import json
    return json.dumps(obj, default=str)

File: node-express/src/controllers/patient.js
const { Router } = require('express');
const pool = require('../config/database');
const { patientToFhir, fhirToPatientFields, searchBundle } = require('../models/fhirMapper');

const router = Router();
const FHIR_JSON = 'application/fhir+json';

router.get('/:id', async (req, res) => {
  const result = await pool.query('SELECT * FROM patients WHERE id = $1', [req.params.id]);
  if (!result.rows[0]) return res.status(404).json({ error: 'Patient not found' });
  res.type(FHIR_JSON).json(patientToFhir(result.rows[0]));
});

router.get('/', async (req, res) => {
  let result;
  if (req.query.identifier) {
    result = await pool.query('SELECT * FROM patients WHERE mrn = $1', [req.query.identifier]);
  } else {
    result = await pool.query('SELECT * FROM patients');
  }
  res.type(FHIR_JSON).json(searchBundle(result.rows.map(patientToFhir)));
});

router.post('/', async (req, res) => {
  const fields = fhirToPatientFields(req.body);
  const result = await pool.query(
    `INSERT INTO patients (family_name, given_name, mrn, birth_date, gender)
     VALUES ($1,$2,$3,$4,$5) RETURNING *`,
    [fields.family_name, fields.given_name, fields.mrn, fields.birth_date, fields.gender]
  );
  const patient = result.rows[0];
  res.status(201).type(FHIR_JSON)
    .set('Location', `/fhir/Patient/${patient.id}`)
    .json(patientToFhir(patient));
});

router.put('/:id', async (req, res) => {
  const fields = fhirToPatientFields(req.body);
  const result = await pool.query(
    `UPDATE patients SET family_name=$1, given_name=$2, mrn=$3, birth_date=$4, gender=$5,
     updated_at=NOW() WHERE id=$6 RETURNING *`,
    [fields.family_name, fields.given_name, fields.mrn, fields.birth_date, fields.gender, req.params.id]
  );
  if (!result.rows[0]) return res.status(404).json({ error: 'Patient not found' });
  res.type(FHIR_JSON).json(patientToFhir(result.rows[0]));
});

module.exports = router;
